libraries/protocol/custom_authorities/restriction_predicate.hxx

```diff
@@ -0,0 +1,595 @@
+   // Compare optional value against comparable type
+   constexpr static bool valid = true;
+   constexpr int8_t operator()(const fc::optional<Field>& f, const Argument& a) const {
+      if (!f.valid()) return predicate_result::Rejection(predicate_result::null_optional);
+       return (*this)(*f, a);
+   }
+};
+// The actual inequality predicates
+template<typename Field, typename Argument> struct predicate_lt : predicate_compare<Field, Argument> {
+   using base = predicate_compare<Field, Argument>;
+   constexpr bool operator()(const Field& f, const Argument& a) const { return base::operator()(f, a) < 0; }
+};
+template<typename Field, typename Argument> struct predicate_le : predicate_compare<Field, Argument> {
+   using base = predicate_compare<Field, Argument>;
+   constexpr bool operator()(const Field& f, const Argument& a) const { return base::operator()(f, a) <= 0; }
+};
+template<typename Field, typename Argument> struct predicate_gt : predicate_compare<Field, Argument> {
+   using base = predicate_compare<Field, Argument>;
+   constexpr bool operator()(const Field& f, const Argument& a) const { return base::operator()(f, a) > 0; }
+};
+template<typename Field, typename Argument> struct predicate_ge : predicate_compare<Field, Argument> {
+   using base = predicate_compare<Field, Argument>;
+   constexpr bool operator()(const Field& f, const Argument& a) const { return base::operator()(f, a) >= 0; }
+};
+
+// Field-in-list predicate
+template<typename F, typename C, typename = void> struct predicate_in : predicate_invalid<F, C> {};
+template<typename Field, typename Element>
+struct predicate_in<Field, flat_set<Element>, std::enable_if_t<comparable_types<Field, Element> && !is_safe<Field>>> {
+   // Simple inclusion check
+   constexpr static bool valid = true;
+   bool operator()(const Field& f, const flat_set<Element>& c) const { return c.count(f) != 0; }
+};
+template<typename Field, typename Element>
+struct predicate_in<fc::safe<Field>, flat_set<Element>, std::enable_if_t<comparable_types<Field, Element>>> {
+   // Check for safe value
+   constexpr static bool valid = true;
+   bool operator()(const fc::safe<Field>& f, const flat_set<Element>& c) const { return c.count(f.value) != 0; }
+};
+template<typename Field, typename Element>
+struct predicate_in<fc::optional<Field>, flat_set<Element>, std::enable_if_t<comparable_types<Field, Element>>> {
+   // Check for optional value
+   constexpr static bool valid = true;
+   bool operator()(const fc::optional<Field>& f, const flat_set<Element>& c) const {
+      if (!f.valid()) return predicate_result::Rejection(predicate_result::null_optional);
+       return c.count(*f) != 0;
+   }
+};
+template<typename Container, typename Element>
+struct predicate_in<Container, flat_set<Element>,
+                    std::enable_if_t<is_container<Container> &&
+                                     comparable_types<typename Container::value_type, Element>>> {
```

Comment:
The `flat_map` type used in `graphene::authorities` is not supported. Perhaps implement something similar to `variant_assertion`.

Reply:
That would be interesting, but I anticipate it'll be a fairly niche use case. I'm calling it future work. :P